### PR TITLE
S4158: Add SE test cases for C# 12 collection expressions

### PR DESCRIPTION
**Pull request by zsolt-kolbay-sonarsource**
Reproduces #8539

### CODE REVIEW

**Review by gregory-paidis-sonarsource — APPROVED** [review]
LGTM

File: analyzers/tests/SonarAnalyzer.Net8.UnitTest/Rules/EmptyCollectionsShouldNotBeEnumeratedTest.cs [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+using ChecksCS = SonarAnalyzer.SymbolicExecution.Roslyn.RuleChecks.CSharp;
+using CS = SonarAnalyzer.Rules.CSharp;
+
+namespace SonarAnalyzer.UnitTest.Rules;
+
+[TestClass]
+public class EmptyCollectionsShouldNotBeEnumeratedTest
```
Comment (pavel-mikula-sonarsource) [inline analyzers/tests/SonarAnalyzer.Net8.UnitTest/Rules/EmptyCollectionsShouldNotBeEnumeratedTest.cs:27]: Adding new file for existing rule is a big red flag. The new UT should be in the existing file.

File: analyzers/tests/SonarAnalyzer.Net8.UnitTest/TestCases/EmptyCollectionsShouldNotBeEnumerated.cs [hunk 2]
```diff
@@ -0,0 +1,29 @@
+﻿using System;
```
Comment (pavel-mikula-sonarsource) [inline analyzers/tests/SonarAnalyzer.Net8.UnitTest/TestCases/EmptyCollectionsShouldNotBeEnumerated.cs:1]: This file has wrong name and lives in a wrong directory.

**Review by pavel-mikula-sonarsource — CHANGES_REQUESTED** [review]
This PR will be affected by the UnitTest->Test rename. 

At the same time, I noticed that both files were wrong.

File: analyzers/tests/SonarAnalyzer.Net8.UnitTest/Rules/EmptyCollectionsShouldNotBeEnumeratedTest.cs (hunk 1 above)
Comment (pavel-mikula-sonarsource) [reply analyzers/tests/SonarAnalyzer.Net8.UnitTest/Rules/EmptyCollectionsShouldNotBeEnumeratedTest.cs:27]: As Greg pointed out, it's .NET 8 that I missed. In that case only the namespace is wrong and should be the same as in other files in the same directory.

..and will need to be updated to `.Test.` after #8544 is done

File: analyzers/tests/SonarAnalyzer.Net8.UnitTest/TestCases/EmptyCollectionsShouldNotBeEnumerated.cs (hunk 2 above)
Comment (pavel-mikula-sonarsource) [reply analyzers/tests/SonarAnalyzer.Net8.UnitTest/TestCases/EmptyCollectionsShouldNotBeEnumerated.cs:1]: ...ah, it's .NET8 so i'ts fine 🤦

**Comment by pavel-mikula-sonarsource** [comment]
The directory rename PR was merged, so you can move the new files now

File: analyzers/tests/SonarAnalyzer.Net8.Test/Rules/EmptyCollectionsShouldNotBeEnumeratedTest.cs [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+[TestClass]
+public class EmptyCollectionsShouldNotBeEnumeratedTest
+{
+    private readonly VerifierBuilder roslynCS = new VerifierBuilder()
+        .AddAnalyzer(() => new CS.SymbolicExecutionRunner())
+        .WithOnlyDiagnostics(ChecksCS.EmptyCollectionsShouldNotBeEnumerated.S4158);
+
+    [TestMethod]
+    public void EmptyCollectionsShouldNotBeEnumerated_Roslyn_CSharp12() =>
+        roslynCS.AddPaths("EmptyCollectionsShouldNotBeEnumerated.cs").WithOptions(ParseOptionsHelper.FromCSharp12).Verify();
```
[the remark this answers is not in the document]
Comment (zsolt-kolbay-sonarsource) [reply analyzers/tests/SonarAnalyzer.Net8.Test/Rules/EmptyCollectionsShouldNotBeEnumeratedTest.cs]: Right, I forgot it doesn't need runtime support. I moved it out of the .NET 8 project.

[the remark this answers is not in the document]
Comment (gregory-paidis-sonarsource) [reply analyzers/tests/SonarAnalyzer.Net8.Test/Rules/EmptyCollectionsShouldNotBeEnumeratedTest.cs]: Ah I missed that too.
The normal UT framework sets `LanguageVersion.Latest`, which is enough to use the new collection expression 🤦

**Review by pavel-mikula-sonarsource — APPROVED** [review]
LGTM

**Pull request merged**